### DISCUSSION TITLE
How to utilise the abstract model from a descriptive model

### DISCUSSION THREAD

**Discussion by ghost** [vttresearch/Predicer#4] [General]
This is to serve the place for discussing the interplay between the abstract model core and a specific, descriptive model.

**Comment by Tasqu** [comment]
# Can we make the descriptive-to-abstract layer reversible for outputs?

While providing easier input to the abstract model core is the primary function of the "descriptive-to-abstract" layer, eventually we'll also need to convert the abstract model output into something descriptive for the users. It would be an amazing asset, if the "descriptive-to-abstract" layer could be made reversible somehow, so that it could also be applied to interpret the results back into something more easily understood.

I have no idea if and how this would work, but I thought it worth writing down at least.

**Reply by ghost** [reply]
Definitely it has to work both ways. The ideas of the abstract core in my opinion are:

- make modelling (creating equations) easier, not to be visible to the average user
- make the model more tight and compact (and thus faster to solve?)

The idea of the descriptive layer is then to be there for the average human user.

**Reply by Tasqu** [reply]
To me, the main idea of the abstract core is to provide a set of basic building blocks and a comprehensive set of equations to constrain the interactions between said building blocks. In principle, it should enable a wide range of functionality without the need of adding new variables or constraints to the model code.

However, depending how complex the desired functionality, the input data required to define the appropriate building blocks and their interactions can get really complex as well. This is where I see the "descriptive models" coming in as sort of pre-made input-output processors for adding certain type of system components _(e.g. specific power plants)_ to the abstract model.

When the user needs to add a system component with new functionality, they would still have to understand the abstract core. However, they could then create a new "descriptive model" for that new system component in order to make creating such system components easier in the future.

**Reply by jkiviluo** [reply]
Yes, agreed. To put a slightly different spin on this, the input data uses methods (for conversion and transfer, but also other functions) that the user can choose. The descriptive-to-abstract layer translates the data and data structures needed by the method (and can also check that all data has been provided). 

Going back to descriptive seems to require those original tags from the descriptive data and combining them with the result variables. In FlexTool it seemed like the easier step and not the same as going the original translation backward.

**Reply by Mastomaki** [reply]
I guess the translation components would have specification how to utilize the abstract model variables to return meaningful results.
